### PR TITLE
Optimistic Lock implementation based on _etag field

### LINKED ISSUE #378: Use Optimistic lock with @Version annotation on _etag field
Hi all,
As said by #372 , how to implement the optimist lock in spring data? I have tried to use in my entity the _etag field (a cosmosdb system field updated automatically in every document changing explained in [cosmos concurrency microsoft guide](https://docs.microsoft.com/bs-latn-ba/azure/cosmos-db/database-transactions-optimistic-concurrency#optimistic-concurrency-control)) with spring data @Version annotation without any result.

### PR DESCRIPTION
**Pull request by dsibilio**
## Fixes #378 
Here you can find a basic implementation of Optimistic Locking via CosmosDB if-match precondition based on the _etag field of an entity.

## Behaviour
Upserts that result in updates with not matching `_etag`s will fail.
Deletes with not matching `_etag`s will fail, but - as of the current implementation - deletes by Id and collection deletion do not apply any concurrency control.

## How-To
The `_etag` field needs to be a `String` field, explicitly declared and marked with the `org.springframework.data.annotation.Version` annotation at field level, inside the entity that needs concurrency control features such as optimistic locking.

E.g.

```java
@Document(collection = "testCollection")
class VersionedVolunteer {
    String id;
    String name;
    @Version
    String _etag;
}
```

## Notes

- A Memoizer has been added to avoid repeating expensive reflection operations to determine if an entity is versioned, what's the id field and eventually in the future more expensive operations such as SpEL expression resolution for certain fields (see #390)
- Two test files have been excluded from Codacy reviews because of the "_etag" field name not matching Java standards.

### CODE REVIEW

**Comment by kushagraThapar** [comment]
@dsibilio can you please resolve the merge conflicts for this PR. I will run additional tests to make sure we are passing everything.

**Comment by dsibilio** [comment]
> @dsibilio can you please resolve the merge conflicts for this PR. I will run additional tests to make sure we are passing everything.

@kushagraThapar conflicts have been merged, ty for the reminder ;-)

File: src/test/java/com/microsoft/azure/spring/data/cosmosdb/repository/support/DocumentDbEntityInformationUnitTest.java [hunk 1]
```diff
@@ -53,16 +57,72 @@ public void testGetCollectionName() {
...
+        @Version
+        private String _etag;
+    }
+    
+    @Data
+    @Document
+    class WrongVersionType {
+        private String id;
+        private String name;
+        private long _etag;
```
[the remark this answers is not in the document]
Comment (dsibilio) [reply src/test/java/com/microsoft/azure/spring/data/cosmosdb/repository/support/DocumentDbEntityInformationUnitTest.java]: This should be solved by [.codacy.yml](https://github.com/microsoft/spring-data-cosmosdb/pull/396/files#diff-c8ab7fb12cd338aaaedb191a0ad93474), why is it still showing?

File: src/test/java/com/microsoft/azure/spring/data/cosmosdb/domain/Person.java [hunk 2]
```diff
@@ -8,18 +8,32 @@
...
+@NoArgsConstructor
 @DocumentIndexingPolicy(includePaths = TestConstants.ORDER_BY_STRING_PATH)
 public class Person {
     private String id;
     private String firstName;
     private String lastName;
     private List<String> hobbies;
     private List<Address> shippingAddresses;
+    @Version
+    private String _etag;
```
[the remark this answers is not in the document]
Comment (dsibilio) [reply src/test/java/com/microsoft/azure/spring/data/cosmosdb/domain/Person.java]: This should be solved by [.codacy.yml](https://github.com/microsoft/spring-data-cosmosdb/pull/396/files#diff-c8ab7fb12cd338aaaedb191a0ad93474), why is it still showing?

File: src/test/java/com/microsoft/azure/spring/data/cosmosdb/repository/support/DocumentDbEntityInformationUnitTest.java [hunk 3]
```diff
@@ -53,16 +57,72 @@ public void testGetCollectionName() {
...
+    } 
+    
+    @Data
+    @Document
+    class VersionOnWrongField {
+        private String id;
+        @Version
+        private String name;
+        private String _etag;
```
[the remark this answers is not in the document]
Comment (dsibilio) [reply src/test/java/com/microsoft/azure/spring/data/cosmosdb/repository/support/DocumentDbEntityInformationUnitTest.java]: This should be solved by [.codacy.yml](https://github.com/microsoft/spring-data-cosmosdb/pull/396/files#diff-c8ab7fb12cd338aaaedb191a0ad93474), why is it still showing?

**Comment by kushagraThapar** [comment]
@dsibilio Can you please resolve conflicts again for this PR, we would like to get it in today if possible. 
We would like to include this in our 2.1.8 release.

**Comment by dsibilio** [comment]
> @dsibilio Can you please resolve conflicts again for this PR, we would like to get it in today if possible.
> We would like to include this in our 2.1.8 release.
> 
> Thanks!

I've merged the conflicts but couldn't run the tests.
I'm having trouble with my internet connection, so if any issues occur feel free to make any amends to my branch - edits should be allowed by any spring-data-cosmosdb mantainer.

**Comment by kushagraThapar** [comment]
> > @dsibilio Can you please resolve conflicts again for this PR, we would like to get it in today if possible.
> > We would like to include this in our 2.1.8 release.
> > Thanks!
> 
> I've merged the conflicts but couldn't run the tests.
> I'm having trouble with my internet connection, so if any issues occur feel free to make any amends to my branch - edits should be allowed by any spring-data-cosmosdb mantainer.

Thanks for fixing the conflicts, I will run the CI for this branch.

**Comment by dsibilio** [comment]
> /azp run

Fixed a JUnit that failed due to checking wrong illegal arguments (those are downstack, checked by another test anyways). You can try again :+1:

**Comment by kushagraThapar** [comment]
@dsibilio, the emulator test keeps failing, I ran it twice to make sure. Can you please fix it ?

**Comment by dsibilio** [comment]
> @dsibilio, the emulator test keeps failing, I ran it twice to make sure. Can you please fix it ?

It seems the "precondition is not met" message has been changed on the master to "pre-condition is not met". I fixed the JUnit test assertion, if you run it again we can verify further errors if any :+1:

**Pull request merged**